experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+}
+
+type codegen struct {
+	includeWKTs           bool
+	includeSourceCodeInfo bool
+
+	currentFile File
+}
+
+func (c *codegen) files(files []File, fds *descriptorpb.FileDescriptorSet) {
+	if c.includeSourceCodeInfo {
+		panic("protocompile/ir: NYI: IncludeSourceCodeInfo()")
+	}
+
+	// Build up all of the imported files. We can't just pull out the transitive
+	// imports for each file because we want the result to be sorted
+	// topologically.
+	for file := range TopoSort(files...) {
+		if !c.includeWKTs && wellknownimports.StandardImport(file.Path()) != "" {
+			continue
+		}
+
+		fdp := new(descriptorpb.FileDescriptorProto)
+		fds.File = append(fds.File, fdp)
+		c.file(file, fdp)
+	}
+}
+
+func (c *codegen) file(file File, fdp *descriptorpb.FileDescriptorProto) {
+	c.currentFile = file
+
+	fdp.Name = addr(file.Path())
+	fdp.Package = addr(file.Context().Package())
+
+	switch file.Syntax() {
+	case syntax.Proto2, syntax.Proto3:
+		fdp.Syntax = addr(file.Syntax().String())
+	case syntax.Edition2023:
+		fdp.Syntax = addr("editions")
+		fdp.Edition = descriptorpb.Edition_EDITION_2023.Enum()
+	}
+
+	for imp := range seq.Values(file.Imports()) {
+		fdp.Dependency = append(fdp.Dependency, imp.Path())
+		if imp.Public {
+			fdp.PublicDependency = append(fdp.PublicDependency, int32(len(fdp.Dependency)-1))
+		}
+		if imp.Weak {
+			fdp.WeakDependency = append(fdp.WeakDependency, int32(len(fdp.Dependency)-1))
+		}
+	}
+
+	for ty := range seq.Values(file.Types()) {
+		if ty.IsEnum() {
+			edp := new(descriptorpb.EnumDescriptorProto)
+			fdp.EnumType = append(fdp.EnumType, edp)
+			c.enum(ty, edp)
+			continue
+		}
+
+		mdp := new(descriptorpb.DescriptorProto)
+		fdp.MessageType = append(fdp.MessageType, mdp)
+		c.message(ty, mdp)
+	}
+
+	// TODO: Services.
+
+	for extn := range seq.Values(file.Extensions()) {
+		fd := new(descriptorpb.FieldDescriptorProto)
+		fdp.Extension = append(fdp.Extension, fd)
+		c.field(extn, fd)
+	}
+
+	for option := range seq.Values(file.Options()) {
+		if fdp.Options == nil {
+			fdp.Options = new(descriptorpb.FileOptions)
+		}
+
+		c.option(option, fdp.Options)
+	}
+}
+
+func (c *codegen) message(ty Type, mdp *descriptorpb.DescriptorProto) {
+	mdp.Name = addr(ty.Name()) // Has a leading dot.
+
+	for field := range seq.Values(ty.Fields()) {
+		fd := new(descriptorpb.FieldDescriptorProto)
+		mdp.Field = append(mdp.Field, fd)
+		c.field(field, fd)
+	}
+
+	for extn := range seq.Values(ty.Extensions()) {
+		fd := new(descriptorpb.FieldDescriptorProto)
+		mdp.Extension = append(mdp.Extension, fd)
+		c.field(extn, fd)
+	}
+
+	for ty := range seq.Values(ty.Nested()) {
+		if ty.IsEnum() {
+			edp := new(descriptorpb.EnumDescriptorProto)
+			mdp.EnumType = append(mdp.EnumType, edp)
+			c.enum(ty, edp)
+			continue
+		}
+
+		mdp := new(descriptorpb.DescriptorProto)
+		mdp.NestedType = append(mdp.NestedType, mdp)
+		c.message(ty, mdp)
+	}
+
+	//nolint:revive // Complains about range_.
```

Comment:
I kind of agree with this linter. You can just name it `extRange` here and `rsvdRange` below, and it helps readability and satisfies the linter.

Reply:
If you insist. :)